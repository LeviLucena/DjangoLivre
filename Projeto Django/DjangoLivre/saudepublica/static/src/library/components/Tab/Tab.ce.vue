<template>
  <div class="br-tab" :class="verifiedClassesActive" :data-counter="counter">
    <nav class="tab-nav" :aria-label="label">
      <ul>
        <li
          v-for="(tab, index) in tabs"
          :key="tab.title"
          class="tab-item"
          :class="{ active: index == selectedIndex }"
          @click="selectTab(index)"
        >
          <template v-if="!tab.icon">
            <button
              v-tooltip="{
                text: tab.tooltipText,
                place: tab.tooltipPlace,
                type: tab.tooltipType,
                timer: tab.tooltipTimer,
              }"
              type="button"
              :data-panel="tab.id"
              :aria-label="tab.ariaLabel ? tab.ariaLabel : tab.id"
              @click="$emit('onTabItemClick', tab)"
            >
              <span class="name">
                {{ tab.title }}
              </span>
            </button>
          </template>

          <template v-if="tab.icon">
            <button
              v-tooltip="{
                text: tab.tooltipText,
                place: tab.tooltipPlace,
                type: tab.tooltipType,
                timer: tab.tooltipTimer,
              }"
              type="button"
              :data-panel="tab.id"
              :aria-label="tab.ariaLabel ? tab.ariaLabel : tab.id"
            >
              <span class="name">
                <span class="d-flex flex-column flex-sm-row">
                  <span class="icon mb-1 mb-sm-0 mr-sm-1">
                    <IconBase :icon-name="tab.icon" />
                  </span>
                  <span class="name">{{ tab.title }}</span>
                </span>
              </span>
            </button>
          </template>
          <span v-if="tab.results !== ''" class="results"> {{ tab.results }} </span>
        </li>
      </ul>
    </nav>
    <div class="tab-content">
      <slot></slot>
    </div>
  </div>
</template>

<script>
import TooltipMixin from "../../mixins/TooltipMixin"
import IconBase from "../IconBase/IconBase.ce.vue"

const BrTab = {
  name: "brTab",
  components: {
    IconBase,
  },
  mixins: [TooltipMixin],
  provide() {
    return {
      parentTab: this,
    }
  },
  props: {
    /** **[OPCIONAL]** Aria-label para descrição da tabela. */
    label: {
      type: String,
      default: "Abas de Navegação Simples",
    },
    /** **[OPCIONAL]** Tipo das densidades do tab as opções válidas são "small", "medium" e "large" */
    density: {
      default: "small",
      type: String,
      validator: function (value) {
        return ["small", "medium", "large"].includes(value)
      },
    },
    /** **[OPCIONAL]** Aplica background escuro e inverte as cores das fontes. */
    inverted: {
      type: Boolean,
      default: false,
    },
    /** **[OPCIONAL]** Indica que o tab apresenta contadores. Se verdadeiro, utilizar o atributo 'results' do subcomponente 'br-tab-item'. */
    counter: {
      type: Boolean,
      default: false,
    },
  },
  emits: [
    /**
     * **[OPCIONAL]** Evento emitido quando oum item do br-tab é clicado.
     *
     * Referência sobre o consumo de eventos:
     * [Eventos Emitidos pelos Componentes](/?path=/story/eventos-emitidos-pelos-componentes--page)
     */
    "onTabItemClick",
  ],
  data() {
    return {
      tabs: [],
      selectedIndex: 0,
      densityProp: this.density,
    }
  },
  computed: {
    verifiedClassesActive() {
      return {
        small: this.densityProp === "small",
        medium: this.densityProp === "medium",
        large: this.densityProp === "large",
        "dark-mode": this.inverted,
      }
    },
  },
  created() {
    this.tabs = this.$children
  },
  mounted() {
    this.$nextTick(function () {
      if (this.tabs) {
        let foundActive = false

        for (let i = 0; i < this.tabs.length; i++) {
          if (this.tabs[i].isActive) {
            this.selectTab(i)
            foundActive = true
            break
          }
        }

        if (!foundActive && this.tabs.length > 0) {
          this.selectTab(0)
        }
      }
    })
  },
  methods: {
    selectTab(i) {
      if (this.tabs) {
        this.selectedIndex = i
        this.tabs.forEach((tab, index) => {
          tab.setActive(index === i)
        })
      }
    },
    addTabItem(tabItem) {
      if (!this.tabs) {
        this.tabs = []
      }
      this.tabs.push(tabItem)
    },
  },
}

export default BrTab
</script>

<style lang="scss">
@import "~@govbr-ds/core/src/partial/scss/_base";
@import "~@govbr-ds/core/src/partial/scss/_utilities";
@import "~@govbr-ds/core/src/components/tab/_tab";
@import url("https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.11.2/css/all.min.css");
</style>
