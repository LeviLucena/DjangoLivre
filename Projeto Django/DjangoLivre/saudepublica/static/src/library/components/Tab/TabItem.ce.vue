<template>
  <div v-show="active" :id="id" class="tab-panel" :class="{ active: active }" :aria-label="ariaLabel">
    <slot></slot>
  </div>
</template>

<script>
import TooltipMixin from "../../mixins/TooltipMixin"

export default {
  name: "TabItem",
  mixins: [TooltipMixin],
  inject: ["parentTab"],
  props: {
    /** **[OPCIONAL]** Título da aba */
    title: {
      type: String,
      default: "",
    },
    /** **[OPCIONAL]** Identificador da aba */
    id: {
      type: String,
      default: "",
    },
    /** **[OPCIONAL]** Ícone que é mostrado na aba */
    icon: {
      type: String,
      default: "",
    },
    /**
     * **[OPCIONAL]** Define o valor do contador.
     * */
    results: {
      type: String,
      default: "",
    },
    /**
     * **[OPCIONAL]** Define se a aba está ativa/aberta.
     * */
    isActive: {
      type: Boolean,
      default: false,
    },
    /**
     * **[OPCIONAL]** Texto descritivo para o atributo `aria-label` (acessibilidade).
     */
    ariaLabel: {
      type: String,
      default: null,
    },
  },
  data() {
    return {
      active: this.isActive,
    }
  },
  mounted() {
    this.parentTab.addTabItem(this)
  },
  methods: {
    setActive(active) {
      this.active = active
    },
  },
}
</script>

<style lang="scss">
@import "~@govbr-ds/core/src/partial/scss/_base";
@import "~@govbr-ds/core/src/components/tab/_tab";

.tab-panel {
  color: var(--text-color);
}
</style>
