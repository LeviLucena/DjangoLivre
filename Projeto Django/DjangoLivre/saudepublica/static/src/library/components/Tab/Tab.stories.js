import { action } from '@storybook/addon-actions'
import { customEventStorybookArgType, kebabiseArgs } from '../../util/Utils.js'
import Tab from './Tab.ce.vue'
import BrTabItem from './TabItem.ce.vue'

const slotDefault = `
  <br-tab-item title="Sobre" id="panel-1" aria-label="Sobre"><p>Sobre</p></br-tab-item>
  <br-tab-item title="Todos" id="panel-2" aria-label="Todos"><p>Todos</p></br-tab-item>
  <br-tab-item title="Notícias" is-active id="panel-3" aria-label="Notícias"><p>Notícias</p></br-tab-item>
  <br-tab-item title="Serviços" id="panel-4" aria-label="Serviços"><p>Serviços</p></br-tab-item>
`

const slotComIcone = `
  <br-tab-item title="Sobre" id="panel-1" icon="home"><p>Sobre</p></br-tab-item>
  <br-tab-item title="Todos" id="panel-2" icon="image"><p>Todos</p></br-tab-item>
  <br-tab-item title="Notícias" id="panel-3" icon="image"><p>Notícias</p></br-tab-item>
  <br-tab-item title="Serviços" id="panel-4" icon="image"><p>Serviços</p></br-tab-item>
`

const slotComContador = `
  <br-tab-item title="Sobre" id="panel-1" results="120">
    <p>Sobre</p>
  </br-tab-item>
  <br-tab-item title="Todos" id="panel-2" results="124">
    <p>Todos</p>
  </br-tab-item>
  <br-tab-item title="Notícias" id="panel-3" results="74">
    <p>Notícias</p>
  </br-tab-item>
  <br-tab-item title="Serviços" id="panel-4" results="16">
    <p>Serviços</p>
  </br-tab-item>
`

const slotSomenteIcone = `
  <br-tab-item id="panel-1" icon="home" tooltip-text="Home"><p>Sobre</p></br-tab-item>
  <br-tab-item id="panel-2" icon="image" tooltip-text="Todos"><p>Todos</p></br-tab-item>
  <br-tab-item id="panel-3" icon="image" tooltip-text="Notícias"><p>Notícias</p></br-tab-item>
  <br-tab-item id="panel-4" icon="image" tooltip-text="Serviços"><p>Serviços</p></br-tab-item>
`

const slotComIconeETooltip = `
  <br-tab-item title="Sobre" id="panel-1" icon="home" tooltip-text="Lorem ipsum dolor sit amet" tooltip-place="bottom"><p>Sobre</p></br-tab-item>
  <br-tab-item title="Todos" id="panel-2" icon="image" ><p>Todos</p></br-tab-item>
  <br-tab-item title="Notícias" id="panel-3" icon="image"><p>Notícias</p></br-tab-item>
  <br-tab-item title="Serviços" id="panel-4" icon="image"><p>Serviços</p></br-tab-item>
`

export default {
  title: 'Componentes/Tab',
  component: Tab,
  subcomponents: {
    'br-tab-item': BrTabItem,
  },
  parameters: {
    controls: {
      exclude: ['tooltipPlace', 'tooltipText', 'tooltipTimer', 'tooltipType'],
    },
    backgrounds: {
      default: 'light',
      values: [
        { name: 'light', value: '#fff' },
        { name: 'dark', value: '#071D41' },
      ],
    },
  },
  argTypes: {
    density: {
      control: { type: 'radio' },
      options: ['small', 'medium', 'large'],
    },
    default: {
      description:
        '**[OBRIGATÓRIO]** Abas do tab, que devem ser nós do tipo **<br-tab-item>** passados por slot. Propriedades possíveis: title (título), id (identificador), icon (ícone), results (valor do contador), isActive (aba ativa/aberta), tooltipText (texto do tooltip), tooltipPlace (top, right, left, bottom), tooltipType (info, warning, error, success), tooltipTimer (tempo em milisegundos)',
      control: 'text',
      type: {
        required: true,
      },
      table: {
        type: {
          summary: 'NodeList of <br-tab-item>',
        },
        defaultValue: {
          summary: '',
        },
      },
    },
    inverted: {
      defaultValue: false,
    },
    label: {
      defaultValue: '',
    },
    onTabItemClick: {
      ...customEventStorybookArgType,
    },
  },
}

const Template = (args) => ({
  props: Object.keys(args),
  setup() {
    return { args: kebabiseArgs(args) }
  },
  template: `<br-tab v-bind="args" @onTabItemClick="handleTabItemClick">${args.default}</br-tab>`,
  methods: { handleTabItemClick: action('onTabItemClick') },
})

const TemplateFundoEscuro = (args) => ({
  props: Object.keys(args),
  setup() {
    return { args: kebabiseArgs(args) }
  },
  template: `
  <div class="bg-gray-60 p-3">
    <br-tab v-bind="args">${args.default}</br-tab>
  </div>`,
})

export const Default = Template.bind({})
Default.args = {
  default: slotDefault,
}

export const ComIcone = Template.bind({})
ComIcone.args = {
  default: slotComIcone,
}

export const ComContador = Template.bind({})
ComContador.args = {
  default: slotComContador,
  counter: true,
}

export const SomenteIcone = Template.bind({})
SomenteIcone.args = {
  default: slotSomenteIcone,
}

export const ComFundoEscuro = TemplateFundoEscuro.bind({})
ComFundoEscuro.args = {
  default: slotDefault,
  inverted: true,
}

export const ComTooltip = Template.bind({})
ComTooltip.args = {
  default: slotComIconeETooltip,
}
