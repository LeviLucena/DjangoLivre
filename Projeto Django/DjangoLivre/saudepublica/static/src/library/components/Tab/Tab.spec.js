import { shallowMount } from '@vue/test-utils'
import Tab from './Tab.ce.vue'
import TabItem from './TabItem.ce.vue'

describe('Tab', () => {
  it('Deve renderizar o componente tab', () => {
    const wrapper = shallowMount(Tab)
    expect(wrapper.classes(`br-tab`)).toBe(true)
  })

  it('Deve exibir o texto da propriedade aria-label do br-tab', () => {
    const wrapper = shallowMount(Tab, {
      propsData: { label: 'Sobre' },
    })
    const tab = wrapper.find('nav[aria-label]').attributes('aria-label')
    expect(tab).toMatch('Sobre')
  })

  it('Deve aplicar a densidade baixa ao br-tab componente', () => {
    const wrapper = shallowMount(Tab, {
      propsData: { density: 'small' },
    })
    expect(wrapper.classes(`small`)).toBe(true)
  })

  it('Deve aplicar o tab invertido', () => {
    const wrapper = shallowMount(Tab, {
      propsData: { inverted: true },
    })
    expect(wrapper.classes('dark-mode')).toBe(true)
  })
})

describe('TabItem', () => {
  it('Deve renderizar o componente tabItem', () => {
    const wrapper = shallowMount(Tab, {
      slots: {
        default: TabItem,
      },
    })
    expect(wrapper.find('.tab-content').exists()).toBe(true)
    expect(wrapper.element).toMatchSnapshot()
  })

  it('Deve alterar a prop title da aba (tabItem)', () => {
    const slotItens = `
      <br-tab-item title="Sobre" id="panel-1" icon="image">
        <p>Sobre</p>
      </br-tab-item>
      <br-tab-item title="Todos" id="panel-2" icon="image">
        <p>Todos</p>
      </br-tab-item>
      <br-tab-item title="Notícias" id="panel-3">
        <p>Notícias</p>
      </br-tab-item>
      <br-tab-item title="Serviços" id="panel-4">
        <p>Serviços</p>
      </br-tab-item>
      <br-tab-item title="Aplicativos" id="panel-5">
        <p>Aplicativos</p>
      </br-tab-item>
      <br-tab-item title="Mídias" id="panel-6">
        <p>Mídias</p>
      </br-tab-item>
    `

    const wrapper = shallowMount(Tab, {
      slots: {
        default: slotItens,
      },
      global: {
        stubs: {
          'br-tab-item': TabItem,
        },
      },
    })
    expect(wrapper.element).toMatchSnapshot()
  })
})
