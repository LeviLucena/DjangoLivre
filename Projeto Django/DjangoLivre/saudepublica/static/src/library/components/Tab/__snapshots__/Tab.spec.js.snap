// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TabItem Deve alterar a prop title da aba (tabItem) 1`] = `
<div
  class="br-tab small"
  data-counter="false"
>
  <nav
    aria-label="Abas de Navegação Simples"
    class="tab-nav"
  >
    <ul>
      
      
    </ul>
  </nav>
  <div
    class="tab-content"
  >
    
    <div
      class="tab-panel"
      id="panel-1"
      style="display: none;"
    >
      
      <p>
        Sobre
      </p>
      
    </div>
    <div
      class="tab-panel"
      id="panel-2"
      style="display: none;"
    >
      
      <p>
        Todos
      </p>
      
    </div>
    <div
      class="tab-panel"
      id="panel-3"
      style="display: none;"
    >
      
      <p>
        Notícias
      </p>
      
    </div>
    <div
      class="tab-panel"
      id="panel-4"
      style="display: none;"
    >
      
      <p>
        Serviços
      </p>
      
    </div>
    <div
      class="tab-panel"
      id="panel-5"
      style="display: none;"
    >
      
      <p>
        Aplicativos
      </p>
      
    </div>
    <div
      class="tab-panel"
      id="panel-6"
      style="display: none;"
    >
      
      <p>
        Mídias
      </p>
      
    </div>
    
  </div>
</div>
`;

exports[`TabItem Deve renderizar o componente tabItem 1`] = `
<div
  class="br-tab small"
  data-counter="false"
>
  <nav
    aria-label="Abas de Navegação Simples"
    class="tab-nav"
  >
    <ul>
      
      
    </ul>
  </nav>
  <div
    class="tab-content"
  >
    
    <tab-item-stub
      icon=""
      id=""
      isactive="false"
      results=""
      title=""
    />
    
  </div>
</div>
`;
